src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,350 @@
+                var argNodeIndex = arg;
+                this._graph.connect(argNodeIndex, 'ret', sgIndex, sgInputs[argIndex].name);
+            } else if (arg.type) {
+                // ioPort - no swizzle
+                this._graph.connect(-1, arg.name, sgIndex, sgInputs[argIndex].name);
+            } else if (arg.node && arg.node.type) {
+                // port node - probably with swizzle
+                this._graph.connect(-1, arg.node.name, sgIndex, sgInputs[argIndex].name, arg.swizzle);
+            } else {
+                if (!arg.port) {
+                    arg = { node: arg.node, port: 'ret', swizzle: arg.swizzle };
+                }
+                // specific port of a node - maybe with swizzle
+                this._graph.connect(arg.node, arg.port, sgIndex, sgInputs[argIndex].name, arg.swizzle);
+            }
+        }
+        return sgIndex;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addOutput
+     * @description Connect a node port to graph material output.
+     * @param {any} arg - Node port to output.
+     * @param {string} type - Type of output.
+     * @param {string} name - Name of output.
+     * @example
+     * builder.addOutput(arg, 'vec3', 'sgVertOff');
+     * builder.addOutput(arg, 'float', 'sgAlpha');
+     * builder.addOutput(arg, 'vec3', 'sgAlbedo');
+     * builder.addOutput(arg, 'vec4', 'sgCustomPass0');
+     */
+    addOutput: function (arg, type, name) {
+        // validate before adding output
+        if (this._addedOutputs[name]) {
+            console.error('pc.ShaderGraphBuilder#addOutput: output ' + name + ' already added!');
+            return;
+        }
+        var valid = this._validateArg(arg, "pc.ShaderGraphBuilder#addNode");
+        if (!valid) {
+            return;
+        }
+
+        // add valid output
+        var ioPort = this._graph.addOutput(type, name);
+
+        this._addedOutputs[name] = ioPort;
+
+        if (typeof(arg) === 'number') {
+            // ret port of a node - no swizzle
+            var argNodeIndex = arg;
+            this._graph.connect(argNodeIndex, 'ret', -1, ioPort.name);
+        } else if (arg.type) {
+            // ioPort - no swizzle
+            this._graph.connect(-1, arg.name, -1, ioPort.name);
+        } else if (arg.node && arg.node.type) {
+            // port node - probably with swizzle
+            this._graph.connect(-1, arg.node.name, -1, ioPort.name, arg.swizzle);
+        } else {
+            if (!arg.port) {
+                arg = { node: arg.node, port: 'ret', swizzle: arg.swizzle };
+            }
+            // specific port of a node - maybe with swizzle
+            this._graph.connect(arg.node, arg.port, -1, ioPort.name, arg.swizzle);
+        }
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#getShaderGraphChunk
+     * @description Get shader graph chunk.
+     * @returns {any} Shader graph chunk.

```

Comment:
should the output type be a real type rather then any?

Reply:
It should be `ShaderGraphNode` - I'll update it

Reply:
updated